cpp/src/arrow/c/abi.h

```diff
@@ -65,6 +65,69 @@ struct ArrowArray {
 
 #endif  // ARROW_C_DATA_INTERFACE
 
+#ifndef ARROW_C_DEVICE_DATA_INTERFACE
+#define ARROW_C_DEVICE_DATA_INTERFACE
+
+// ArrowDeviceType is compatible with dlpack DLDeviceType for portability
+// it uses the same values for each enum as the equivalent kDL<type> from dlpack.h
+#ifdef __cplusplus
+typedef enum : int32_t {
+#else
+typedef enum {
+#endif
+  // CPU device, same as using ArrowArray directly
+  kArrowCPU = 1,
+  // CUDA GPU Device
+  kArrowCUDA = 2,
+  // Pinned CUDA CPU memory by cudaMallocHost
+  kArrowCUDAHost = 3,
+  // OpenCL Device
+  kArrowOpenCL = 4,
+  // Vulkan buffer for next-gen graphics
+  kArrowVulkan = 7,
+  // Metal for Apple GPU
+  kArrowMetal = 8,
+  // Verilog simulator buffer
+  kArrowVPI = 9,
+  // ROCm GPUs for AMD GPUs
+  kArrowROCM = 10,
+  // Pinned ROCm CPU memory allocated by hipMallocHost
+  kArrowROCMHost = 11,
+  // Reserved for extension
+  // used to quickly test extension devices,
+  // semantics can differ based on the implementation
+  kArrowExtDev = 12,
+  // CUDA managed/unified memory allocated by cudaMallocManaged
+  kArrowCUDAManaged = 13,
+  // unified shared memory allocated on a oneAPI non-partitioned
+  // device. call to oneAPI runtime is required to determine the
+  // device type, the USM allocation type and the sycl context it
+  // is bound to
+  kArrowOneAPI = 14,
+  // GPU support for next-gen WebGPU standard
+  kArrowWebGPU = 15,
+  // Qualcomm Hexagon DSP
+  kArrowHexagon = 16,
+} ArrowDeviceType;
+
+struct ArrowDeviceArray {
+  // the private_date and release callback of the arrow array
+  // should contain any necessary information and structures
+  // related to freeing the array according to the device it
+  // is allocated on, rather than having a separate release
+  // callback embedded here.
+  struct ArrowArray* array;
+  int device_id;
```

Comment:
Is an `int` sufficient, or do we want to be future-proof and make this `int64_t`?

Reply:
I'll future proof and set it to an `int64_t` just in case.

Reply:
CUDA (NVIDIA GPUs) and ROCm (AMD GPUs) both use `int` to specify a device ID in their APIs.

SYCL uses a `cl_device_id` which is an opaque type where I'm not sure if an `int32_t` or `int64_t` is sufficient to capture it. Maybe @aregm could have someone help guide here? 

dlpack has historically used an `int` / `int32_t` for 6+ years.

Reply:
anyone have objections to using `int64_t` for future-proof reasons? Should I just use `int32_t` instead?

Reply:
+1 for `int64_t`. We don't know what future device APIs may choose here.

Reply:
> CUDA (NVIDIA GPUs) and ROCm (AMD GPUs) both use `int` to specify a device ID in their APIs.
> 
> SYCL uses a `cl_device_id` which is an opaque type where I'm not sure if an `int32_t` or `int64_t` is sufficient to capture it. Maybe @aregm could have someone help guide here?
> 
> dlpack has historically used an `int` / `int32_t` for 6+ years.

DPC++ uses level zero (https://github.com/oneapi-src/level-zero/tree/master) as the primary C API. As to the device ID it is stored as `uint32_t` currently (see https://spec.oneapi.io/level-zero/latest/core/api.html#_CPPv422ze_device_properties_t) and it is not expected to exceed `int64_t`, so +1 as well.

Reply:
Once remark, though - the assumption is that Arrow will not go beyond a single host, anyone who decided to go low-level knows how to map [UUID](https://spec.oneapi.io/level-zero/latest/core/api.html#ze-uuid-t) to host/device. But with the upcoming CXL capability, this can be a problem. Let's leave it int64 and address the issue when it comes. So +1.